examples/example_customdns_test.go

```diff
@@ -0,0 +1,165 @@
+// Copyright (C) MongoDB, Inc. 2022-present.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+package examples
+
+import (
+	"context"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/miekg/dns"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+func resolve(ctx context.Context, cache *dnsCache, in *dns.Conn, out *dns.Conn) {
+	for ctx.Err() == nil {
+		q, err := in.ReadMsg()
+		if err != nil {
+			// TODO: Handle error.
+			continue
+		}
+		if len(q.Question) != 1 {
+			// Multiple questions in a single query is not actually used in real life.
+			continue
+		}
+
+		a, err := func() (*dns.Msg, error) {
+			cache.lock.Lock()
+			defer cache.lock.Unlock()
+
+			now := time.Now()
+			if rr, ok := cache.records[q.Question[0]]; ok && rr.exp.After(now) {
+				a := new(dns.Msg)
+				a.SetReply(q)
+				a.Compress = false
+				a.Answer = append(a.Answer, rr.record)
+				return a, nil
+			}
+
+			err := out.WriteMsg(q)
```

Comment:
I agree we need a finer granularity of locking for better performance. The lock during server response is to prevent multiple DNS queries when the record is expired. Otherwise, the caching misses its goal of "limiting DNS queries". As an example, I think a total lockdown is clear enough to indicate "concurrency accessibility" yet understandable enough.

Reply:
I think, as an example, we don't need _optimal_ locking. I was going to suggest a `sync.RWMutex` instead of a `sync.Mutex`, and potentially read-locking around the block above and write-locking around the block below, but that might be overkill.

> the caching misses its goal of "limiting DNS queries"

To clarify, do you mean that one `resolve` goroutine could release the lock after discovering an expired query, start querying DNS before locking again, and another `resolve` goroutine could start the same DNS query?

Reply:
Correct. Every new connection attempt starts a `resolve` routine from `Dial: func(ctx context.Context, network, address string) (net.Conn, error)`. Therefore, without a lock, `resolve` may simultaneously emit multiple DNS queries for the same or different name.
I also tried `sync.RWMutex`. However, the RWMutex in Golang lacks the capability of upgrading from a read lock to a write lock because Golang considers it an anti-pattern. In this way, RWMutex still has the issue mentioned above.
Go1.18 introduced `TryLock` for mutexes, which helps to resolve the multi-entry (though still a bad use case of mutexes in the Golang philosophy) like:
```
for {
    rwMutex.RLock()
    // read logic
    rwMutex.RUnlock()
    if !expired {
        break
    } else if rwMutex.TryLock() {
        // emit DNS query
        // update cache
        rwMutex.Unlock()
        break
    }
}
```
I'm afraid it is too rambling in an example of the "custom dialer".

Reply:
> However, the RWMutex in Golang lacks the capability of upgrading from a read lock to a write lock because Golang considers it an anti-pattern.

Ah TIL 🤔 . That all makes sense to me; I think the example is good as-is, and users can modify the concurrency and error logging as they see fit.

Reply:
Could we avoid locking if we used a Go channel?

```go
// start this "worker" wherever you initialize the "dialer"
go func(cache *dnsCache) {
    for question := range questionQueue {
        // cache record
    }
}(cache)

// enqueue questions to cache within the resolver
questionQueue <- q
```

Reply:
If we are going the have a worker, the worker works in a "single producer multiple consumer" pattern in my opinion considering multiple connections may initialize name queries. On the other hand, connections may query different hostnames, for example in a cluster, if I understand correctly. In this case, I'm not sure how a single "producer" channel sends out answers to corresponding "consumer" connections. The Go channel works more like a broadcast rather than a landline in my understanding. Though it can be resolved by having a channel per hostname, still the locking alone is too complicated in such an example.

Reply:
The locks are also specified at the constructor level, right? If you specify the worker at that level, shouldn't the results be the same?

```go
func NewDialer() dialer {
	cache := &dnsCache{
		records: make(map[dns.Question]*RR),
		lock:    &sync.Mutex{}, // same level as the worker
	}
	
        questionQueue  := make(QuestionType)
        go func(cache *dnsCache) {
            for question := range questionQueue {
                // cache record, should be FIFO
            }
        }(cache)
        
	return dialer{
		Dialer: &net.Dialer{
			Resolver: &net.Resolver{
				PreferGo: true,
				Dial: func(ctx context.Context, network, address string) (net.Conn, error) {
					outConn, err := net.Dial(network, address)
					conn, inConn := net.Pipe()
					if err == nil {
						go resolve(ctx, cache, questionQueue, &dns.Conn{Conn: inConn}, &dns.Conn{Conn: outConn})
					}
					return conn, err
				},
			},
		},
		cache: cache,
	}
}
```